src/app/clusters/operational-credentials-server/operational-credentials-cluster.cpp

```diff
@@ -941,448 +688,719 @@ bool emberAfOperationalCredentialsClusterUpdateNOCCallback(app::CommandHandler *
+
+    // Try to send the queued events as soon as possible for this fabric. If the just emitted leave event won't
+    // be sent this time, it will likely not be delivered at all for the following reasons:
+    // - removing the fabric expires all associated ReadHandlers, so all subscriptions to
+    //   the leave event will be cancelled.
+    // - removing the fabric removes all associated access control entries, so generating
+    //   subsequent reports containing the leave event will fail the access control check.
+    mContext->interactionContext.eventsGenerator.ScheduleUrgentEventDeliverySync(fabricIndex);
+}
 
-    CHIP_ERROR err = fabricTable.SignVIDVerificationRequest(commandData.fabricIndex, commandData.clientChallenge,
-                                                            attestationChallenge, responseData);
-    if (err == CHIP_ERROR_INVALID_ARGUMENT)
-    {
-        commandObj->AddStatus(commandPath, Status::ConstraintError);
-        return true;
-    }
+void OperationalCredentialsCluster::OnFabricRemoved(const FabricTable & fabricTable, FabricIndex fabricIndex)
+{
+    ChipLogDetail(Zcl, "OpCreds: Fabric index 0x%x was removed", static_cast<unsigned>(fabricIndex));
 
-    if (err != CHIP_NO_ERROR)
-    {
-        // We have no idea what happened; just report failure.
-        StatusIB status(err);
-        commandObj->AddStatus(commandPath, status.mStatus);
-        return true;
-    }
+    // We need to withdraw the advertisement for the now-removed fabric, so need
+    // to restart advertising altogether.
+    GetDNSSDServer().StartServer();
 
-    Commands::SignVIDVerificationResponse::Type response;
-    response.fabricIndex          = responseData.fabricIndex;
-    response.fabricBindingVersion = responseData.fabricBindingVersion;
-    response.signature            = responseData.signature.Span();
-    commandObj->AddResponse(commandPath, response);
+    EventManagement::GetInstance().FabricRemoved(fabricIndex);
 
-    return true;
+    NotifyAttributeChanged(OperationalCredentials::Attributes::CommissionedFabrics::Id);
+    NotifyAttributeChanged(OperationalCredentials::Attributes::Fabrics::Id);
+}
+
+void OperationalCredentialsCluster::OnFabricUpdated(const FabricTable & fabricTable, FabricIndex fabricIndex)
+{
+    NotifyAttributeChanged(OperationalCredentials::Attributes::CommissionedFabrics::Id);
+    NotifyAttributeChanged(OperationalCredentials::Attributes::Fabrics::Id);
+}
+
+void OperationalCredentialsCluster::OperationalCredentialsNotifyAttribute(AttributeId attributeId)
+{
+    NotifyAttributeChanged(attributeId);
+}
+
+FabricTable & OperationalCredentialsCluster::GetFabricTable()
+{
+    return mFabricTable;
+}
+
+FailSafeContext & OperationalCredentialsCluster::GetFailSafeContext()
+{
+    return mFailSafeContext;
+}
+
+Credentials::DeviceAttestationCredentialsProvider * OperationalCredentialsCluster::GetDACProvider()
+{
+    return Credentials::GetDeviceAttestationCredentialsProvider();
```

Comment:
Add a TODO with an issue number that the global instances should be removed and we should instead use injection in a followup PR for all these.